go.mod

```diff
@@ -32,6 +32,7 @@ require (
 	github.com/google/go-cmp v0.6.0
 	github.com/google/go-jsonnet v0.20.0
 	github.com/google/shlex v0.0.0-20191202100458-e7afc7fbc510
```

Comment:
@Matovidlo (cc @jachym-tousek-keboola ) you are mixing together different network stack layers :apple: :pear: 

`udp/tcp/kcp` - are low-level protocols to transfer `bytes`

`rpc/grcp/http` - are serialization/API layers

You can run `grpc` over `kcp`, example: https://github.com/Lynnworld/grpc-kcp-transport

I selected `net/rpc` + `gob` (it will be a next PR) as serialization/API layer.

https://github.com/cockroachdb/rpc-bench

Reply:
1. I understand that they are `L4` layer. However do we really need such a complication? 
2. You did not respond on streaming part, yet we can optimize it that much but do we need that ? We can have the best implementation vs good enough to fit the project needs.
3. Benchmark is 4 years old, so not relevant now.
4. I don't like the choice as gRPC has over 20k stars and is very relevant these days for internal communication.
I understand that it is not worth to analyse/rewrite it now. I just wanted to point out that this is quite a decision and I am not happy about that.

Reply:
> 1. I understand that they are L4 layer. However do we really need such a complication?

:one: TCP/KCP
- Tunning of TCP protocol may be complicated, and it must be done on the OS level, [read more](https://blog.cloudflare.com/optimizing-tcp-for-high-throughput-and-low-latency).
- This also applies to the UDP protocol, but there are much fewer settings at the OS level.
- KCP can be well tunned in app level.
- On my `localhost`, TCP is 3x-4x faster than my disk, so it can actually be a bottleneck.
  - Will we run into it in a week, a month, never? I don't know.
  - Well, I'm sure you'll be happy to have already something to experiment with there.
  - Replacing these low-level things later, if you don't think about it id advance, can be quite a problem and time-consuming.
- In the [next PR](https://github.com/keboola/keboola-as-code/pull/1895), I set TCP as default protocol for initial testing.

:two: yamux - connections multiplexing
- Establishing a connection has a a overhead and delay.
- A separate connection for each slice would consume more resources.
- You could simply run into a connections count limit.

> 2. You did not respond on streaming part, yet we can optimize it that much but do we need that ? We can have the best implementation vs good enough to fit the project needs.

I would like to answer, but I don't know how and what exactly you mean. Our entire solution is based on streaming, the protocol has no effect on it.

gRPC supports streaming of entities, it is great, but we are sending chunks of bytes, it is possible in each protocol.

> 3. Benchmark is 4 years old, so not relevant now.

The link to the benchmark was there to show that they are similar solutions, also looked at various other benchmarks, depending on the use-case one or the other will be faster, but there is no significant difference.

> 4. I don't like the choice as gRPC has over 20k stars and is very relevant these days for internal communication.
I understand that it is not worth to analyse/rewrite it now. I just wanted to point out that this is quite a decision and I am not happy about that.

`net/rpc`  +`encoding/gob` are parts of the standard library, but yes, gRPC is great (for API/Server2Server comm.).

I chosed `net/rpc` because gRPC seemed overkill for our simple use-case.